File: src/horizon_mcp/client/http.py
"""Async HTTP client for Horizon API with CSRF, auth injection, and safe retries."""

from __future__ import annotations

import asyncio
import logging
import time
import uuid
from typing import Any

import httpx
from tenacity import (
    retry,
    retry_if_exception_type,
    stop_after_attempt,
    wait_exponential,
)

from horizon_mcp.auth.base import AuthProvider
from horizon_mcp.client.errors import HorizonError, parse_error_response
from horizon_mcp.settings import HorizonSettings

logger = logging.getLogger("horizon_mcp.client")

# PUT/DELETE endpoints verified as idempotent  -  initially empty, populated incrementally
RETRYABLE_ENDPOINTS: set[tuple[str, str]] = set()

# Status codes that warrant retry on safe methods
_RETRYABLE_STATUS = frozenset({429, 502, 503, 504})


class HorizonClient:
    """Async HTTP client for Horizon API.

    Features:
    - Auth header injection via AuthProvider
    - CSRF token handling (fetch + refresh on 403)
    - Auto-retry for GET/HEAD only (3x exponential backoff, 429/5xx/connection)
    - No retry for POST/PATCH by default
    - PUT/DELETE retry only for endpoints in RETRYABLE_ENDPOINTS allowlist
    - CSRF 403 single retry on any method (no mutation occurred)
    - Request correlation via request_id (UUID4)
    """

    def __init__(self, settings: HorizonSettings, auth: AuthProvider) -> None:
        self._settings = settings
        self._auth = auth
        self._csrf_token: str | None = None
        self._initialized = False
        self._init_lock = asyncio.Lock()
        client_kw = auth.client_kwargs()
        self._http = httpx.AsyncClient(
            base_url=settings.base_url,
            verify=client_kw.pop("verify", settings.verify_ssl),
            timeout=httpx.Timeout(settings.timeout, read=settings.timeout),
            follow_redirects=True,
            **client_kw,
        )

    # -- Public API ---------------------------------------------------------

    async def get(
        self, path: str, params: dict[str, Any] | None = None, **kwargs: Any
    ) -> dict[str, Any]:
        return await self._request_json("GET", path, params=params, **kwargs)

    async def post(
        self, path: str, json: Any = None, **kwargs: Any
    ) -> dict[str, Any]:
        return await self._request_json("POST", path, json=json, **kwargs)

    async def put(
        self, path: str, json: Any = None, **kwargs: Any
    ) -> dict[str, Any]:
        return await self._request_json("PUT", path, json=json, **kwargs)

    async def patch(
        self, path: str, json: Any = None, **kwargs: Any
    ) -> dict[str, Any]:
        return await self._request_json("PATCH", path, json=json, **kwargs)

    async def delete(self, path: str, **kwargs: Any) -> dict[str, Any] | None:
        resp = await self._request("DELETE", path, **kwargs)
        if resp.status_code == 204:
            return None
        return resp.json()

    async def get_bytes(
        self, path: str, params: dict[str, Any] | None = None, **kwargs: Any
    ) -> bytes:
        resp = await self._request("GET", path, params=params, **kwargs)
        return resp.content

    async def get_text(
        self, path: str, params: dict[str, Any] | None = None, **kwargs: Any
    ) -> str:
        resp = await self._request("GET", path, params=params, **kwargs)
        return resp.text

    async def close(self) -> None:
        await self._http.aclose()

    # -- CSRF ---------------------------------------------------------------

    async def fetch_csrf_token(self) -> str | None:
        """Fetch CSRF token from Horizon.

        Checks the auth provider first (OIDC captures csrf-token during
        browser login), then tries the JSON API endpoint, then falls back
        to reading the ``csrf-token`` cookie set by Play Framework.
        Returns None if no source provides a token  -  in that case the
        client will send ``Csrf-Token: nocheck`` on mutating requests.
        """
        # Auth provider may have captured csrf-token (e.g., OIDC browser flow)
        provider_token = self._auth.csrf_token
        if provider_token:
            self._csrf_token = provider_token
            return self._csrf_token

        # Ensure auth session is ready before making HTTP calls
        await self._auth.refresh_if_needed()

        # Re-check after refresh  -  OIDC captures csrf-token during browser flow
        provider_token = self._auth.csrf_token
        if provider_token:
            self._csrf_token = provider_token
            return self._csrf_token

        try:
            resp = await self._http.get(
                "/api/v1/security/csrf",
                headers=await self._auth.get_headers(),
            )
            if resp.status_code == 200:
                data = resp.json()
                self._csrf_token = data.get("token") or data.get("csrfToken")
                if self._csrf_token:
                    return self._csrf_token
        except Exception:
            logger.debug("CSRF JSON endpoint unavailable  -  checking cookies")

        # Fallback: Play Framework sets a csrf-token cookie
        cookie_token = self._http.cookies.get("csrf-token")
        if cookie_token:
            self._csrf_token = cookie_token
            return self._csrf_token

        return None

    # -- Lazy initialization ------------------------------------------------

    async def _ensure_initialized(self) -> None:
        """Lazy init: auth + CSRF + whoami on first request."""
        if self._initialized:
            return
        async with self._init_lock:
            if self._initialized:
                return  # Another coroutine already initialized
            await self._do_lazy_init()
            self._initialized = True

    async def _do_lazy_init(self) -> None:
        """Trigger auth, fetch CSRF token, and run whoami."""
        from horizon_mcp.client.state import set_horizon_version, set_principal_name

        # Trigger auth (browser login for Play Session, no-op for API key/mTLS)
        await self._auth.refresh_if_needed()

        # CSRF token
        await self.fetch_csrf_token()

        # Whoami  -  use raw httpx to avoid recursion through _request
        try:
            headers = await self._auth.get_headers()
            resp = await self._http.get(
                "/api/v1/security/principals/self", headers=headers
            )
            if resp.status_code == 200:
                principal = resp.json()
                identity = principal.get("identity") or {}
                name = (
                    identity.get("identifier")
                    or principal.get("identifier")
                    or principal.get("name")
                    or "unknown"
                )
                set_principal_name(name)

                version = principal.get("_horizonVersion")
                set_horizon_version(version)
                logger.info(
                    "Authenticated as: %s (Horizon %s)",
                    name,
                    version or "unknown",
                )
            else:
                logger.warning(
                    "Whoami returned %d  -  continuing without principal info",
                    resp.status_code,
                )
        except Exception as exc:
            logger.warning("Whoami failed: %s  -  continuing", exc)

    # -- Internal -----------------------------------------------------------

    async def _request(
        self,
        method: str,
        path: str,
        *,
        request_id: str | None = None,
        timeout_override: float | None = None,
        reauth_attempted: bool = False,
        **kwargs: Any,
    ) -> httpx.Response:
        request_id = request_id or uuid.uuid4().hex[:12]
        start = time.monotonic()

        await self._ensure_initialized()
        await self._auth.refresh_if_needed()
        headers = await self._auth.get_headers()
        headers["X-Request-ID"] = request_id

        # CSRF handling for mutating methods (Play Framework)
        if method.upper() not in ("GET", "HEAD"):
            if self._csrf_token:
                headers["Csrf-Token"] = self._csrf_token
            else:
                # Play Framework API convention: "nocheck" bypasses CSRF for
                # non-browser clients authenticated via API keys.
                headers["Csrf-Token"] = "nocheck"

        if timeout_override:
            kwargs["timeout"] = httpx.Timeout(timeout_override)

        try:
            resp = await self._do_request(method, path, headers, request_id, **kwargs)
        except httpx.ConnectError as exc:
            logger.error(
                "Connection failed",
                extra={"request_id": request_id, "method": method, "path": path},
            )
            raise HorizonError(
                status_code=0,
                message=f"Connection to {self._settings.base_url} failed: {exc}",
                remediation="Check HORIZON_URL and network connectivity.",
            ) from exc

        duration_ms = int((time.monotonic() - start) * 1000)

        logger.info(
            "HTTP %s %s → %d (%dms)",
            method, path, resp.status_code, duration_ms,
            extra={"request_id": request_id, "method": method, "path": path,
                   "status": resp.status_code, "duration_ms": duration_ms},
        )

        # CSRF 403 → single retry after token refresh
        if resp.status_code == 403 and self._is_csrf_rejection(resp):
            logger.info("CSRF rejected  -  refreshing token and retrying",
                        extra={"request_id": request_id})
            await self.fetch_csrf_token()
            headers["Csrf-Token"] = self._csrf_token or "nocheck"
            resp = await self._do_request(method, path, headers, request_id, **kwargs)
            if resp.status_code >= 400:
                raise parse_error_response(resp.status_code, resp.content)
            return resp

        # Auth failure retry: 401 or non-CSRF 403 → re-authenticate once
        if resp.status_code in (401, 403) and not reauth_attempted:
            logger.info(
                "Auth rejected (%d)  -  attempting re-authentication",
                resp.status_code,
                extra={"request_id": request_id},
            )
            await self._auth.mark_auth_failed()
            await self._auth.refresh_if_needed()
            headers = await self._auth.get_headers()
            headers["X-Request-ID"] = request_id
            if method.upper() not in ("GET", "HEAD"):
                headers["Csrf-Token"] = self._csrf_token or "nocheck"
            try:
                resp = await self._do_request(
                    method, path, headers, request_id, **kwargs
                )
            except httpx.ConnectError as exc:
                raise HorizonError(
                    status_code=0,
                    message=f"Connection to {self._settings.base_url} failed: {exc}",
                    remediation="Check HORIZON_URL and network connectivity.",
                ) from exc
            if resp.status_code >= 400:
                raise parse_error_response(resp.status_code, resp.content)
            return resp

        if resp.status_code >= 400:
            raise parse_error_response(resp.status_code, resp.content)

        return resp

    async def _do_request(
        self,
        method: str,
        path: str,
        headers: dict[str, str],
        request_id: str,
        **kwargs: Any,
    ) -> httpx.Response:
        """Execute HTTP request with retry logic based on method safety."""
        upper = method.upper()

        # Safe methods: auto-retry
        if upper in ("GET", "HEAD"):
            return await self._retry_request(method, path, headers=headers, **kwargs)

        # PUT/DELETE: retry only if on the verified allowlist
        if upper in ("PUT", "DELETE") and (upper, path) in RETRYABLE_ENDPOINTS:
            return await self._retry_request(method, path, headers=headers, **kwargs)

        # POST/PATCH and non-allowlisted PUT/DELETE: no retry
        return await self._http.request(method, path, headers=headers, **kwargs)

    @retry(
        retry=retry_if_exception_type((httpx.ConnectError, httpx.ReadTimeout)),
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=1, min=1, max=10),
        reraise=True,
    )
    async def _retry_request(
        self, method: str, path: str, **kwargs: Any
    ) -> httpx.Response:
        """Execute request with tenacity retry for transient failures."""
        resp = await self._http.request(method, path, **kwargs)
        if resp.status_code in _RETRYABLE_STATUS:
            raise httpx.ReadTimeout(
                f"Retryable status {resp.status_code}",
                request=resp.request,
            )
        return resp

    async def _request_json(
        self, method: str, path: str, **kwargs: Any
    ) -> dict[str, Any]:
        resp = await self._request(method, path, **kwargs)
        if not resp.content:
            return {}
        return resp.json()

    @staticmethod
    def _is_csrf_rejection(resp: httpx.Response) -> bool:
        """Check if a 403 is specifically a CSRF rejection vs. auth/perm issue."""
        try:
            body = resp.json()
            raw_error = body.get("error", "")
            error_str = raw_error if isinstance(raw_error, str) else str(raw_error)
            message = body.get("message", "")
            message = message if isinstance(message, str) else str(message)
            return "csrf" in error_str.lower() or "csrf" in message.lower()
        except Exception:
            return False

File: src/client/http.ts
import { randomUUID } from 'node:crypto';
import {
  Agent,
  FormData as UndiciFormData,
  fetch as undiciFetch,
} from 'undici';
import type { RequestInit as UndiciRequestInit } from 'undici';
import type { ZodType } from 'zod';

import type { AuthProvider } from '../auth/base.js';
import { getLogger } from '../logging.js';
import {
  RETRYABLE_ENDPOINTS,
  connectionCauseCode,
  isConnectionError,
  positiveSeconds,
  readJsonBounded,
  versionCompatibilityLog,
} from './client-helpers.js';
import {
  HorizonCsrfError,
  HorizonError,
  HorizonResponseValidationError,
  parseErrorResponse,
} from './errors.js';
import { composeWithTimeout } from './request-signal.js';
import { withRetry } from './retry.js';

const logger = getLogger('horizon_mcp.client');

// Rate-limit boundaries for the auto re-auth path on 401/403.
const REAUTH_MIN_INTERVAL_MS = 5 * 60 * 1000;
const REAUTH_MAX_INTERVAL_MS = 30 * 60 * 1000;

/**
 * Common per-request options accepted by the public verb helpers.
 * `schema` enables opt-in Zod validation of the parsed body.
 * `allowCsrfNoCheck` is a per-request escape hatch for endpoints that
 * historically tolerate the literal `Csrf-Token: nocheck`.
 */
export interface RequestOptions<T = unknown> {
  timeout?: number;
  schema?: ZodType<T>;
  allowCsrfNoCheck?: boolean;
}

export interface MultipartPart {
  fieldName: string;
  filename: string;
  mimeType: string;
  data: Buffer | string;
}

export interface HorizonClientOptions {
  timeout?: number;
  exportTimeout?: number;
  verifySsl: boolean;
  testedVersions?: readonly string[];
  warnVersions?: readonly string[];
  onAuthReject?: () => void;
}

export class HorizonClient {
  private readonly _baseUrl: string;
  private readonly _auth: AuthProvider;
  private readonly _timeout: number;
  /** CSV export request budget in seconds, usable directly as the timeout request option. */
  readonly exportTimeout: number;
  private readonly _agent: Agent;
  private readonly _testedVersions: readonly string[];
  private readonly _warnVersions: readonly string[];
  private readonly _onAuthReject?: () => void;
  private _csrfToken: string | undefined;
  private _initialized = false;
  private _initPromise: Promise<void> | null = null;
  private _authRejectNoted = false;

  // Rate-limit state for the 401 -> re-auth path.
  private _lastReauthAt: number | null = null;
  private _reauthBackoffMs: number = REAUTH_MIN_INTERVAL_MS;

  // Captured during lazy init
  principalName: string | undefined;
  horizonVersion: string | undefined;

  constructor(
    baseUrl: string,
    auth: AuthProvider,
    options: HorizonClientOptions,
  ) {
    this._baseUrl = baseUrl.replace(/\/+$/, '');
    this._auth = auth;
    this._timeout = positiveSeconds('timeout', options.timeout, 30) * 1000;
    this.exportTimeout = positiveSeconds(
      'exportTimeout',
      options.exportTimeout,
      120,
    );
    this._testedVersions = options.testedVersions ?? [];
    this._warnVersions = options.warnVersions ?? [];
    this._onAuthReject = options.onAuthReject;

    // Build undici Agent with TLS connect options
    const authConnect = auth.getDispatcherOptions();
    const connectOptions: Agent.Options['connect'] = {
      ...((typeof authConnect === 'object' ? authConnect : {}) as Record<
        string,
        unknown
      >),
      rejectUnauthorized: options.verifySsl,
    };
    this._agent = new Agent({ connect: connectOptions });

    if (!options.verifySsl) {
      logger.warning(
        'HORIZON_VERIFY_SSL is disabled - TLS certificate verification is OFF (insecure; use only against trusted test hosts)',
      );
    }
  }

  // -- Public API -----------------------------------------------------------

  async get<T = unknown>(
    path: string,
    params?: URLSearchParams,
    opts?: RequestOptions<T>,
  ): Promise<T> {
    const url = params ? `${path}?${params.toString()}` : path;
    return this._requestJson<T>('GET', url, undefined, opts);
  }

  async post<T = unknown>(
    path: string,
    body?: unknown,
    opts?: RequestOptions<T>,
  ): Promise<T> {
    return this._requestJson<T>(
      'POST',
      path,
      {
        body: body !== undefined ? JSON.stringify(body) : undefined,
        timeoutMs: opts?.timeout ? opts.timeout * 1000 : undefined,
        allowCsrfNoCheck: opts?.allowCsrfNoCheck,
      },
      opts,
    );
  }

  async put<T = unknown>(
    path: string,
    body: unknown,
    opts?: RequestOptions<T>,
  ): Promise<T> {
    return this._requestJson<T>(
      'PUT',
      path,
      {
        body: JSON.stringify(body),
        allowCsrfNoCheck: opts?.allowCsrfNoCheck,
      },
      opts,
    );
  }

  async patch<T = unknown>(
    path: string,
    body: unknown,
    opts?: RequestOptions<T>,
  ): Promise<T> {
    return this._requestJson<T>(
      'PATCH',
      path,
      {
        body: JSON.stringify(body),
        allowCsrfNoCheck: opts?.allowCsrfNoCheck,
      },
      opts,
    );
  }

  async delete(
    path: string,
    opts?: RequestOptions<unknown>,
  ): Promise<unknown | null> {
    const resp = await this._request('DELETE', path, {
      allowCsrfNoCheck: opts?.allowCsrfNoCheck,
    });
    if (resp.status === 204) return null;
    const parsed = await readJsonBounded<unknown>(resp, path);
    if (opts?.schema) {
      return this._validateOrThrow(opts.schema, parsed, path, resp.status);
    }
    return parsed;
  }

  /**
   * DELETE with a JSON request body. Some Horizon endpoints (e.g. role/team
   * member removal) take a JSON array in the DELETE body. Mirrors `delete`
   * otherwise (204 -> null, else parsed JSON).
   */
  async deleteWithBody(
    path: string,
    body: unknown,
    opts?: RequestOptions<unknown>,
  ): Promise<unknown | null> {
    const resp = await this._request('DELETE', path, {
      body: JSON.stringify(body),
      allowCsrfNoCheck: opts?.allowCsrfNoCheck,
    });
    if (resp.status === 204) return null;
    const parsed = await readJsonBounded<unknown>(resp, path);
    if (opts?.schema) {
      return this._validateOrThrow(opts.schema, parsed, path, resp.status);
    }
    return parsed;
  }

  async getBytes(path: string): Promise<ArrayBuffer> {
    const resp = await this._request('GET', path);
    return resp.arrayBuffer();
  }

  async getText(path: string): Promise<string> {
    const resp = await this._request('GET', path);
    return resp.text();
  }

  async postText(
    path: string,
    body?: unknown,
    opts?: RequestOptions<string>,
  ): Promise<string> {
    const resp = await this._request('POST', path, {
      body: body !== undefined ? JSON.stringify(body) : undefined,
      timeoutMs: opts?.timeout ? opts.timeout * 1000 : undefined,
      allowCsrfNoCheck: opts?.allowCsrfNoCheck,
    });
    return resp.text();
  }

  async postMultipart<T = unknown>(
    path: string,
    parts: MultipartPart[],
  ): Promise<T> {
    const formData = new UndiciFormData();
    for (const part of parts) {
      const blobPart =
        typeof part.data === 'string'
          ? part.data
          : Uint8Array.from(part.data).buffer;
      const blob = new Blob([blobPart], { type: part.mimeType });
      formData.append(part.fieldName, blob, part.filename);
    }

    await this._ensureInitialized();
    const headers = await this._buildHeaders('POST');
    // Remove content-type - let fetch set it with boundary
    delete headers['Content-Type'];

    const requestId = randomUUID().slice(0, 12);
    headers['X-Request-Id'] = requestId;

    const start = performance.now();
    const resp = await undiciFetch(`${this._baseUrl}${path}`, {
      method: 'POST',
      headers,
      body: formData,
      dispatcher: this._agent,
      signal: composeWithTimeout(this._timeout),
    } as UndiciRequestInit);

    const durationMs = Math.round(performance.now() - start);
    logger.info(`HTTP POST ${path} -> ${resp.status} (${durationMs}ms)`, {
      request_id: requestId,
      method: 'POST',
      path,
      status: resp.status,
      duration_ms: durationMs,
    });

    if (resp.status >= 400) {
      throw parseErrorResponse(resp.status, await resp.text());
    }

    return (await readJsonBounded<T>(resp, path)) as T;
  }

  async request(
    method: string,
    path: string,
    opts?: { timeout?: number },
  ): Promise<Response> {
    return this._request(method, path, {
      timeoutMs: opts?.timeout ? opts.timeout * 1000 : undefined,
    });
  }

  async close(): Promise<void> {
    await this._agent.close();
  }

  // -- CSRF -----------------------------------------------------------------

  async fetchCsrfToken(): Promise<string | undefined> {
    // Step 1: Check auth provider (pre-refresh)
    let providerToken = this._auth.csrfToken;
    if (providerToken) {
      this._csrfToken = providerToken;
      return this._csrfToken;
    }

    // Step 2: Trigger auth refresh
    await this._auth.refreshIfNeeded();

    // Step 3: Re-check after refresh (OIDC captures csrf-token during browser flow)
    providerToken = this._auth.csrfToken;
    if (providerToken) {
      this._csrfToken = providerToken;
      return this._csrfToken;
    }

    // Step 4: Try JSON API endpoint
    try {
      const headers = await this._auth.getHeaders();
      const resp = await undiciFetch(`${this._baseUrl}/api/v1/security/csrf`, {
        method: 'GET',
        headers,
        dispatcher: this._agent,
        signal: composeWithTimeout(this._timeout),
      });
      if (resp.status === 200) {
        const data = (await resp.json()) as Record<string, unknown>;
        const token =
          (data['token'] as string | undefined) ??
          (data['csrfToken'] as string | undefined);
        if (token) {
          this._csrfToken = token;
          return this._csrfToken;
        }
      }
    } catch (err) {
      logger.debug(`CSRF JSON endpoint unavailable - checking cookies: ${err}`);
    }

    // Step 5: Fallback to csrf-token cookie from response headers
    // Note: undici.fetch doesn't auto-store cookies like httpx. The cookie
    // would have been captured by the Play Session auth provider instead.

    // Step 6: No CSRF source - will send "nocheck"
    return undefined;
  }

  // -- Lazy initialization --------------------------------------------------

  private async _ensureInitialized(): Promise<void> {
    if (this._initialized) return;
    if (!this._initPromise) {
      this._initPromise = this._doLazyInit();
    }
    await this._initPromise;
  }

  private async _doLazyInit(): Promise<void> {
    await this._performInit(false);
    this._initialized = true;
  }

  /**
   * Strict, eager variant of lazy init for the HTTP per-session flow. Runs the
   * exact same path as lazy init (refresh -> CSRF -> whoami -> capture
   * principal/version -> version-compat log) but THROWS on any failure instead
   * of logging-and-continuing, and marks the client initialized so the first
   * tool call does not re-run init (and mutating calls still carry the captured
   * CSRF token). The thrown error never echoes the caller's credential.
   */
  async validateAuth(): Promise<void> {
    await this._performInit(true);
    this._initialized = true;
  }

  /**
   * Shared init body for lazy (`strict=false`, log-and-continue) and eager
   * `validateAuth` (`strict=true`, throw). Factored into one method so the two
   * paths cannot drift.
   */
  private async _performInit(strict: boolean): Promise<void> {
    // 1. Trigger auth (no-op for API key / mTLS / cert-forward).
    await this._auth.refreshIfNeeded();

    // 2. Fetch CSRF token.
    await this.fetchCsrfToken();

    // 3. Whoami - capture principal name + Horizon version.
    const headers = await this._auth.getHeaders();
    if (this._csrfToken) {
      headers['Csrf-Token'] = this._csrfToken;
    }

    let resp: Response;
    try {
      resp = await undiciFetch(
        `${this._baseUrl}/api/v1/security/principals/self`,
        {
          method: 'GET',
          headers,
          dispatcher: this._agent,
          signal: composeWithTimeout(this._timeout),
        },
      );
    } catch (err) {
      if (strict) {
        throw new HorizonError(0, {
          message: `Authentication check could not reach Horizon: ${err}`,
          remediation: 'Check HORIZON_URL and network connectivity.',
        });
      }
      logger.warning(`Whoami failed: ${err} - continuing`);
      return;
    }

    if (resp.status !== 200) {
      if (strict) {
        // parseErrorResponse redacts secrets; the whoami body never contains
        // the caller's credential.
        const text = await resp.text().catch(() => '');
        throw parseErrorResponse(resp.status, text);
      }
      logger.warning(
        `Whoami returned ${resp.status} - continuing without principal info`,
      );
      return;
    }

    const principal = (await resp.json()) as Record<string, unknown>;
    const identity = (principal['identity'] ?? {}) as Record<string, unknown>;
    this.principalName =
      (identity['identifier'] as string | undefined) ??
      (principal['identifier'] as string | undefined) ??
      (principal['name'] as string | undefined) ??
      'unknown';
    this.horizonVersion = principal['_horizonVersion'] as string | undefined;

    logger.info(
      `Authenticated as: ${this.principalName} (Horizon ${this.horizonVersion ?? 'unknown'})`,
    );

    // 4. Log version compatibility.
    if (this.horizonVersion) {
      const compatibility = versionCompatibilityLog(
        this.horizonVersion,
        this._testedVersions,
        this._warnVersions,
      );
      if (compatibility) {
        logger[compatibility.level](compatibility.message);
      }
    }

    if (!strict) this._auth.markValidated();
  }

  // -- Internal request pipeline --------------------------------------------

  private async _request(
    method: string,
    path: string,
    opts?: {
      body?: string;
      timeoutMs?: number;
      reauthAttempted?: boolean;
      allowCsrfNoCheck?: boolean;
    },
  ): Promise<Response> {
    const requestId = randomUUID().slice(0, 12);
    const start = performance.now();
    const timeoutMs = opts?.timeoutMs ?? this._timeout;

    await this._ensureInitialized();
    const headers = await this._buildHeaders(method, opts?.allowCsrfNoCheck);
    headers['X-Request-Id'] = requestId;

    const fetchOpts: UndiciRequestInit & { dispatcher: Agent } = {
      method,
      headers,
      dispatcher: this._agent,
      signal: composeWithTimeout(timeoutMs),
    };
    if (opts?.body) {
      fetchOpts.body = opts.body;
      headers['Content-Type'] = 'application/json';
    }

    const fullUrl = `${this._baseUrl}${path}`;
    let resp: Response;

    try {
      resp = await this._doRequest(method, fullUrl, fetchOpts, path);
    } catch (err) {
      const causeCode = connectionCauseCode(err);
      if (isConnectionError(err)) {
        throw new HorizonError(0, {
          message: `Connection to ${this._baseUrl} failed${
            causeCode ? ` (${causeCode})` : ''
          }: ${err}`,
          remediation: 'Check HORIZON_URL and network connectivity.',
        });
      }
      throw err;
    }

    const durationMs = Math.round(performance.now() - start);
    logger.info(`HTTP ${method} ${path} -> ${resp.status} (${durationMs}ms)`, {
      request_id: requestId,
      method,
      path,
      status: resp.status,
      duration_ms: durationMs,
    });

    // CSRF 403 -> single retry after token refresh.
    // Fail closed: if we cannot acquire a new token, throw HorizonCsrfError
    // rather than silently sending Csrf-Token: nocheck.
    if (resp.status === 403 && (await this._isCsrfRejection(resp.clone()))) {
      logger.info('CSRF rejected - refreshing token and retrying', {
        request_id: requestId,
      });
      await this.fetchCsrfToken();
      if (!this._csrfToken) {
        if (opts?.allowCsrfNoCheck) {
          headers['Csrf-Token'] = 'nocheck';
        } else {
          throw new HorizonCsrfError(
            'Failed to acquire CSRF token after refresh',
            {
              detail: `${method} ${path}`,
            },
          );
        }
      } else {
        headers['Csrf-Token'] = this._csrfToken;
      }
      fetchOpts.headers = headers;
      fetchOpts.signal = composeWithTimeout(timeoutMs);
      resp = await undiciFetch(fullUrl, fetchOpts);
      if (resp.status >= 400) {
        throw parseErrorResponse(resp.status, await resp.text());
      }
      return resp;
    }

    // Auth failure retry: 401 or 403 -> re-authenticate once, but rate-limit
    // the attempt so a wedged auth provider cannot hammer the IdP.
    if (
      (resp.status === 401 || resp.status === 403) &&
      !opts?.reauthAttempted
    ) {
      const now = Date.now();
      const elapsed =
        this._lastReauthAt === null ? Infinity : now - this._lastReauthAt;

      if (elapsed < this._reauthBackoffMs) {
        logger.warning(
          `Auth rejected (${resp.status}) - skipping re-auth (last attempt ${Math.round(
            elapsed / 1000,
          )}s ago, floor ${Math.round(this._reauthBackoffMs / 1000)}s)`,
          { request_id: requestId },
        );
        // Return the 401/403 to the caller for normal error handling.
        if (resp.status >= 400) {
          this._noteAuthReject();
          throw parseErrorResponse(resp.status, await resp.text());
        }
        return resp;
      }

      logger.info(
        `Auth rejected (${resp.status}) - attempting re-authentication`,
        { request_id: requestId },
      );
      // Exponential backoff: if we re-authed recently (within the previous
      // window), the next floor doubles - capped at REAUTH_MAX_INTERVAL_MS.
      if (this._lastReauthAt !== null && elapsed < REAUTH_MAX_INTERVAL_MS * 2) {
        this._reauthBackoffMs = Math.min(
          this._reauthBackoffMs * 2,
          REAUTH_MAX_INTERVAL_MS,
        );
      } else {
        this._reauthBackoffMs = REAUTH_MIN_INTERVAL_MS;
      }
      this._lastReauthAt = now;

      await this._auth.markAuthFailed();
      await this._auth.refreshIfNeeded();
      return this._request(method, path, {
        ...opts,
        reauthAttempted: true,
      });
    }

    if (resp.status >= 400) {
      if (resp.status === 401 || resp.status === 403) this._noteAuthReject();
      throw parseErrorResponse(resp.status, await resp.text());
    }

    return resp;
  }

  private _noteAuthReject(): void {
    if (this._authRejectNoted) return;
    this._authRejectNoted = true;
    try {
      this._onAuthReject?.();
    } catch {
      // The upstream authentication error remains the surfaced failure.
    }
  }

  private async _doRequest(
    method: string,
    url: string,
    fetchOpts: UndiciRequestInit & { dispatcher: Agent },
    path: string,
  ): Promise<Response> {
    const upper = method.toUpperCase();

    // Safe methods: auto-retry
    if (upper === 'GET' || upper === 'HEAD') {
      return withRetry(() =>
        undiciFetch(url, {
          ...fetchOpts,
          signal: composeWithTimeout(this._timeout),
        }),
      );
    }

    // PUT/DELETE: retry only if on the verified allowlist.
    // RETRYABLE_ENDPOINTS is intentionally empty - idempotent-write retry is
    // disabled, so this branch is currently inert by design (see line 22).
    if (
      (upper === 'PUT' || upper === 'DELETE') &&
      RETRYABLE_ENDPOINTS.has(`${upper}:${path}`)
    ) {
      return withRetry(() =>
        undiciFetch(url, {
          ...fetchOpts,
          signal: composeWithTimeout(this._timeout),
        }),
      );
    }

    // POST/PATCH and non-allowlisted: no retry
    return undiciFetch(url, fetchOpts);
  }

  private async _buildHeaders(
    method: string,
    allowCsrfNoCheck?: boolean,
  ): Promise<Record<string, string>> {
    await this._auth.refreshIfNeeded();
    const headers = await this._auth.getHeaders();

    // CSRF handling for mutating methods.
    // If no token is available we only fall back to the legacy
    // `nocheck` literal when the caller explicitly opted in.
    if (method.toUpperCase() !== 'GET' && method.toUpperCase() !== 'HEAD') {
      if (this._csrfToken) {
        headers['Csrf-Token'] = this._csrfToken;
      } else if (allowCsrfNoCheck) {
        headers['Csrf-Token'] = 'nocheck';
      }
      // Otherwise: omit the header and let the server decide.
    }

    return headers;
  }

  private async _requestJson<T>(
    method: string,
    path: string,
    opts?: {
      body?: string;
      timeoutMs?: number;
      allowCsrfNoCheck?: boolean;
    },
    reqOpts?: RequestOptions<T>,
  ): Promise<T> {
    const resp = await this._request(method, path, opts);
    const parsed = await readJsonBounded<T>(resp, path);
    if (reqOpts?.schema) {
      return this._validateOrThrow(
        reqOpts.schema,
        parsed,
        path,
        resp.status,
      ) as T;
    }
    return parsed as T;
  }

  private _validateOrThrow<T>(
    schema: ZodType<T>,
    value: unknown,
    path: string,
    statusCode: number,
  ): T {
    const result = schema.safeParse(value);
    if (!result.success) {
      const issues = result.error.issues
        .map((issue) => `${issue.path.join('.') || '<root>'}: ${issue.message}`)
        .join('; ');
      throw new HorizonResponseValidationError({
        path,
        statusCode,
        issues,
      });
    }
    return result.data;
  }

  private async _isCsrfRejection(resp: Response): Promise<boolean> {
    try {
      const body = (await resp.json()) as Record<string, unknown>;
      const rawError = body['error'] ?? '';
      const errorStr =
        typeof rawError === 'string' ? rawError : String(rawError);
      const message = body['message'] ?? '';
      const messageStr =
        typeof message === 'string' ? message : String(message);
      return (
        errorStr.toLowerCase().includes('csrf') ||
        messageStr.toLowerCase().includes('csrf')
      );
    } catch (err) {
      logger.debug(`CSRF rejection check failed to parse body: ${err}`);
      return false;
    }
  }
}
